File: browser/src/js/moneysocket/layer/consumer.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const Layer = require("./layer.js").Layer;
const ConsumerNexus = require("../nexus/consumer.js").ConsumerNexus;


class ConsumerLayer extends Layer {
    constructor() {
        super();
        this.onproviderinfo = null;
        this.onping = null;
    }

    setupConsumerNexus(below_nexus) {
        var n = new ConsumerNexus(below_nexus, this);
        n.onproviderinfo = (function(nexus, msg) {
            this.onProviderInfo(nexus, msg);
        }).bind(this);
        n.onping = (function(nexus, msecs) {
            this.onPing(nexus, msecs);
        }.bind(this));
        return n;
    }

    announceNexus(below_nexus) {
        console.log("consumer layer got nexus, starting handshake");
        var consumer_nexus = this.setupConsumerNexus(below_nexus);
        this._trackNexus(consumer_nexus, below_nexus);

        consumer_nexus.startHandshake(this.consumerFinishedCb.bind(this));
    }

    consumerFinishedCb(consumer_nexus) {
        this._trackNexusAnnounced(consumer_nexus);
        this.sendLayerEvent(consumer_nexus, "NEXUS_ANNOUNCED");
        if (this.onannounce != null) {
            this.onannounce(consumer_nexus);
        }
        consumer_nexus.startPinging();
    }

    revokeNexus(below_nexus) {
        var consumer_nexus = this.nexuses[
            this.nexus_by_below[below_nexus.uuid]];
        super.revokeNexus(below_nexus);
        consumer_nexus.stopPinging();
    }

    ///////////////////////////////////////////////////////////////////////////

    onProviderInfo(consumer_nexus, msg) {
        if (this.onproviderinfo != null) {
            this.onproviderinfo(consumer_nexus, msg);
        }
    }

    onPing(consumer_nexus, msecs) {
        if (this.onping != null) {
            this.onping(consumer_nexus, msecs);
        }
    }
}


exports.ConsumerLayer = ConsumerLayer;

File: python/moneysocket/layer/consumer.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

from moneysocket.nexus.consumer import ConsumerNexus
from moneysocket.layer.layer import Layer


class ConsumerLayer(Layer):
    def __init__(self):
        super().__init__()
        self.onping = None
        self.waiting_for_app = {}

    def announce_nexus(self, below_nexus):
        consumer_nexus = self.setup_consumer_nexus(below_nexus)
        self._track_nexus(consumer_nexus, below_nexus)
        consumer_nexus.start_handshake(self.consumer_finished_cb)

    def setup_consumer_nexus(self, below_nexus):
        n = ConsumerNexus(below_nexus, self)
        n.onping = self.on_ping
        return n

    def revoke_nexus(self, below_nexus):
        consumer_nexus = self.nexuses[self.nexus_by_below[below_nexus.uuid]]
        super().revoke_nexus(below_nexus)
        consumer_nexus.stop_pinging()

    def consumer_finished_cb(self, consumer_nexus):
        self._track_nexus_announced(consumer_nexus)
        self.send_layer_event(consumer_nexus, "NEXUS_ANNOUNCED")
        if self.onannounce:
            self.onannounce(consumer_nexus)

    def on_ping(self, consumer_nexus, msecs):
        if self.onping:
            self.onping(consumer_nexus, msecs)
